lib/tools/broccoli/builder.ts

```diff
@@ -5,29 +5,66 @@ let broccoli = require('broccoli');
-import destCopy = require('./node-modules-dest-copy');
+import destCopyLib = require('./node-modules-dest-copy');
+
+var gulp = require("gulp");
+var vinylFilterSince = require('vinyl-filter-since')
+var through = require("through2"); 
 
 export class Builder implements IBroccoliBuilder {
+   private nodeModules: any = {};
+  
   constructor(private $fs: IFileSystem,
     private $nodeModulesTree: INodeModulesTree,
     private $projectDataService: IProjectDataService,
     private $logger: ILogger) { }
   
-  public prepareNodeModules(absoluteOutputPath: string, projectDir: string): IFuture<void> {
+  public prepareNodeModules(absoluteOutputPath: string, projectDir: string, lastModifiedTime?: Date): IFuture<void> {
     return (() => {
-      // TODO: figure out a better way for doing this
-      this.$projectDataService.initialize(projectDir);
-      let cachedNodeModulesPath = this.$projectDataService.getValue("node-modules-cache-path").wait();
-      if (cachedNodeModulesPath && this.$fs.exists(cachedNodeModulesPath).wait()) {
-        let diffTree = new TreeDiffer(cachedNodeModulesPath);
-        let diffTreeResult = diffTree.diffTree(path.join(projectDir, absoluteOutputPath, "node_modules"));
-        
-        if(diffTreeResult.changedDirectories.length > 0 || diffTreeResult.removedDirectories.length > 0) {
-          this.rebuildNodeModulesTree(absoluteOutputPath, projectDir).wait();
-        }
-      } else { 
-         this.rebuildNodeModulesTree(absoluteOutputPath, projectDir).wait();
+      let nodeModulesPath = path.join(projectDir, "node_modules");
+            
+      if(lastModifiedTime) {
+        let pipeline = gulp.src(path.join(projectDir, "node_modules/**"))
+          .pipe(vinylFilterSince(lastModifiedTime))
+          .pipe(through.obj( (chunk: any, enc: any, cb: Function) => {
+              let rootModuleName = chunk.path.split(nodeModulesPath)[1].split(path.sep)[1];
+              if(rootModuleName) {
+                let rootModuleFullPath = path.join(nodeModulesPath, rootModuleName);
+                this.nodeModules[rootModuleFullPath] = rootModuleFullPath;
+              }
+              
+              cb(null); 
+          }))
+          .pipe(gulp.dest(absoluteOutputPath));
+          
+         let future = new Future<void>();
+         
+         pipeline.on('end', (err: any, data: any) => {
+           if(err) {
+             future.throw(err);
+           } else {
+            future.return();
+           }
+         });
+         
+         future.wait();
+      } else {
+         let nodeModulesDirectories = this.$fs.readDirectory(nodeModulesPath).wait();
+         _.each(nodeModulesDirectories, nodeModuleDirectoryName => {
+           let nodeModuleFullPath = path.join(nodeModulesPath, nodeModuleDirectoryName);
+           this.nodeModules[nodeModuleFullPath] = nodeModuleFullPath;
+         })
       }
+      
+     let destCopy = $injector.resolve(destCopyLib.DestCopy, {
```

Comment:
isn't this the global injector? Maybe you should pass the injector in the constructor and use it from there.